### PR TITLE
Revert "Revert "[Doc] [Jobs] Document `JobDetails`, `JobType` and `DriverInfo`"" (#30500)

### PR DESCRIPTION
**Pull request by architkulkarni**
Cherry picks #30500 to the release branch.

Closes https://github.com/ray-project/ray/issues/30550. Verified the failure locally and verified that this PR fixes it.

An integration test that runs jobs with a conda `runtime_env` was failing on the release branch, likely due to a downstream dependency change. The error message we see is:

```
(raylet) Traceback (most recent call last):
(raylet)   File "/Users/archit/ray/python/ray/_private/workers/default_worker.py", line 8, in <module>
(raylet)     import ray
(raylet)   File "/Users/archit/ray/python/ray/__init__.py", line 163, in <module>
(raylet)     from ray import autoscaler  # noqa: E402,F401
(raylet)   File "/Users/archit/ray/python/ray/autoscaler/__init__.py", line 1, in <module>
(raylet)     from ray.autoscaler import sdk
(raylet)   File "/Users/archit/ray/python/ray/autoscaler/sdk/__init__.py", line 1, in <module>
(raylet)     from ray.autoscaler.sdk.sdk import (
(raylet)   File "/Users/archit/ray/python/ray/autoscaler/sdk/sdk.py", line 9, in <module>
(raylet)     from ray.autoscaler._private import commands
(raylet)   File "/Users/archit/ray/python/ray/autoscaler/_private/commands.py", line 23, in <module>
(raylet)     from ray.autoscaler._private.autoscaler import AutoscalerSummary
(raylet)   File "/Users/archit/ray/python/ray/autoscaler/_private/autoscaler.py", line 16, in <module>
(raylet)     from six.moves import queue
(raylet) ModuleNotFoundError: No module named 'six'
```

It's passing on master due to a PR #30500  which adds `ray[default]` to the installation for the conda tests in CI.  As a side effect, this likely installs the correct dependency `six` due to it appearing somewhere in the dependency tree.  This PR is a cherry pick of that PR to the release branch.

My hypothesis is that there was recently a breaking downstream dependency that was hidden on master (because this PR was already merged in master) but caused the release branch to break.

Note: This is a fix for CI only, because when testing the runtime_env `conda` feature on commits that don't have built wheels, we have to manually install Ray's dependencies in the conda environment before linking to the Ray source from that commit.  The feature itself doesn't go through this codepath; the codepath for released versions is tested during release tests.

This PR also contains a minor jobs doc change, but rather than split it out into a new PR, I'll just cherry pick the existing PR to save time for the release.

Original context of this PR reproduced below:

---
Reverts #30499. The reason this PR was originally reverted because it caused Jobs runtime env tests to fail, because runtime_env CI tests were installing "ray" into the runtime_env conda environment instead of "ray[default]", which is required for Jobs. This PR fixes this by installing "ray[default]" in runtime_env CI tests.

Original PR description:

The classes JobDetails, JobType and DriverInfo are exposed to the user because they are the return type of JobSubmissionClient.get_job_info, so we need to document them.

This PR adds them to the API reference.

## Related issue number

Closes https://github.com/ray-project/ray/issues/30550

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@xwjiang2010 This PR just fixes a CI test issue, it's not a problem with the feature. The argument for picking it is so the release manager doesn't have to scan through the failed test and check that the failure is the known failure (`test_http_job_server.py::test_job_submit[conda_yaml]`). On the other hand, if it slows the release too much, it's not necessary to pick it.

**Comment by xwjiang2010** [comment]
Thank you @architkulkarni. I prefer to pick what is needed for the tests to pass.
However, I will leave it to serve TLs to decide if it's low risk enough to cherry-pick the whole thing or have a minimal fix instead.

**Review by simon-mo — APPROVED** [review]
Look like the doc changes are necessary anyway.

**Comment by xwjiang2010** [comment]
Ok, sounds good. I am going to merge this together with https://github.com/ray-project/ray/pull/30801. And I will monitor branch build to see if tests are fixed.

**Pull request merged**
